source/extensions/filters/network/kafka/kafka_request.h

```diff
@@ -0,0 +1,398 @@
+
+protected:
+  RequestContextSharedPtr context_;
+  BT buffer_;
+};
+
+template <typename RT, typename BT>
+ParseResponse BufferedParser<RT, BT>::parse(const char*& buffer, uint64_t& remaining) {
+  context_->remaining_request_size_ -= buffer_.feed(buffer, remaining);
+  if (buffer_.ready()) {
+    // after a successful parse, there should be nothing left
+    ASSERT(0 == context_->remaining_request_size_);
+    RT request = buffer_.get();
+    request.header() = context_->request_header_;
+    ENVOY_LOG(trace, "parsed request {}: {}", *context_, request);
+    MessageSharedPtr msg = std::make_shared<RT>(request);
+    return ParseResponse::parsedMessage(msg);
+  } else {
+    return ParseResponse::stillWaiting();
+  }
+}
+
+// names of Buffers/Parsers are influenced by org.apache.kafka.common.protocol.Protocol names
+
+#define DEFINE_REQUEST_PARSER(REQUEST_TYPE, VERSION)                                               \
+  class REQUEST_TYPE##VERSION##Parser                                                              \
+      : public BufferedParser<REQUEST_TYPE, REQUEST_TYPE##VERSION##Buffer> {                       \
+  public:                                                                                          \
+    REQUEST_TYPE##VERSION##Parser(RequestContextSharedPtr ctx) : BufferedParser{ctx} {};           \
+  };
+
+// === ABSTRACT REQUEST ========================================================
+
+class Request : public Message {
+public:
+  /**
+   * Request header fields need to be initialized by user in case of newly created requests
+   */
+  Request(INT16 api_key) : request_header_{api_key, 0, 0, ""} {};
+
+  Request(const RequestHeader& request_header) : request_header_{request_header} {};
+
+  RequestHeader& header() { return request_header_; }
+
+  INT16& apiVersion() { return request_header_.api_version_; }
+  INT16 apiVersion() const { return request_header_.api_version_; }
+
+  INT32& correlationId() { return request_header_.correlation_id_; }
+
+  NULLABLE_STRING& clientId() { return request_header_.client_id_; }
+
+  size_t encode(Buffer::Instance& dst, EncodingContext& encoder) const {
+    size_t written{0};
+    written += encoder.encode(request_header_.api_key_, dst);
+    written += encoder.encode(request_header_.api_version_, dst);
+    written += encoder.encode(request_header_.correlation_id_, dst);
+    written += encoder.encode(request_header_.client_id_, dst);
+    written += encodeDetails(dst, encoder);
+    return written;
+  }
+
+  std::ostream& print(std::ostream& os) const override final {
+    os << request_header_ << " "; // not very pretty
+    return printDetails(os);
+  }
+
+protected:
+  virtual size_t encodeDetails(Buffer::Instance&, EncodingContext&) const PURE;
+
+  virtual std::ostream& printDetails(std::ostream&) const PURE;
+
+  RequestHeader request_header_;
+};
+
+// === OFFSET COMMIT (8) =======================================================
```

Comment:
In general, I think this code would benefit from starting to split it into multiple files. Can we potentially have a file for base objects/classes, and then a file per message type since they seem to be fairly large? I think that would make the code easier to read and review as it grows.

Reply:
naming them `request_type.h` and putting them in `messages` subdirectory
I envision one header will contain both request & response for the same type